Sources/NIOHTTP2/HTTP2ChannelHandler+InlineStreamMultiplexer.swift

```diff
@@ -0,0 +1,205 @@
+    /// independent stream of HTTP/2 frames. This multiplexer achieves this by creating a
+    /// number of in-memory `HTTP2StreamChannel` objects, one for each stream. These operate
+    /// on ``HTTP2Frame/FramePayload`` objects as their base communication
+    /// atom, as opposed to the regular NIO `SelectableChannel` objects which use `ByteBuffer`
+    /// and `IOData`.
+    public struct StreamMultiplexer {
+        private let inlineStreamMultiplexer: InlineStreamMultiplexer
+
+        /// Cannot be created by users.
+        internal init(_ inlineStreamMultiplexer: InlineStreamMultiplexer) {
+            self.inlineStreamMultiplexer = inlineStreamMultiplexer
+        }
+
+        /// Create a new `Channel` for a new stream initiated by this peer.
+        ///
+        /// This method is intended for situations where the NIO application is initiating the stream. For clients,
+        /// this is for all request streams. For servers, this is for pushed streams.
+        ///
+        /// > Note: Resources for the stream will be freed after it has been closed.
+        ///
+        /// - parameters:
+        ///     - promise: An `EventLoopPromise` that will be succeeded with the new activated channel, or
+        ///         failed if an error occurs.
+        ///     - streamStateInitializer: A callback that will be invoked to allow you to configure the
+        ///         `ChannelPipeline` for the newly created channel.
+        public func createStreamChannel(promise: EventLoopPromise<Channel>?, _ streamStateInitializer: @escaping StreamInitializer) {
+            self.inlineStreamMultiplexer.createStreamChannel(promise: promise, streamStateInitializer)
+        }
+
+        /// Create a new `Channel` for a new stream initiated by this peer.
+        ///
+        /// This method is intended for situations where the NIO application is initiating the stream. For clients,
+        /// this is for all request streams. For servers, this is for pushed streams.
+        ///
+        /// > Note: Resources for the stream will be freed after it has been closed.
+        ///
+        /// - Parameter streamStateInitializer: A callback that will be invoked to allow you to configure the
+        ///         `ChannelPipeline` for the newly created channel.
+        /// - Returns: An `EventLoopFuture` containing the created `Channel`, fulfilled after the supplied `streamStateInitializer` has been executed on it.
+        public func createStreamChannel(_ streamStateInitializer: @escaping StreamInitializer) -> EventLoopFuture<Channel> {
+            self.inlineStreamMultiplexer.createStreamChannel(streamStateInitializer)
+        }
+    }
+
+    /// A ``StreamMultiplexer`` which can be used to create new outbound HTTP/2 streams.
+    ///
+    /// > Note: The ``NIOHTTP2Handler`` must have been initialized with a ``StreamConfiguration``
+    /// for this value to be non-nil.
+    public var multiplexer: StreamMultiplexer? {
```

Comment:
Can we document that this value is only non-`nil` after the handler has been added to the pipeline?

Reply:
This is slightly awkward as we don't have an `EventLoop` on `init`. Adding one to the new inits is fine but we can't add them to the existing inits without breaking API. Rick and I discussed a couple of options. These are the most promising ones:

One is to have a future-based API which bangs the event loop since it will only ever be `nil` when using the old `init` (where we will never have a `StreamMultiplexer`).

The other is having a promise-based API which fails the promise if `self.eventLoop` is `nil`.

We think the first is fine since it'll only fail if the old soon-to-be-deprecated `init` is used and we can document it appropriately. WDYT @Lukasa?

Reply:
I've gone ahead and implemented the first one as a basis of discussion.

Reply:
I think that's fine by me. We may want to keep a non-future-based one as well, perhaps behind a `preconditionOnEventLoop` flag or similar.